src/commands/trace/trace.ts

```diff
@@ -0,0 +1,147 @@
+import chalk from 'chalk'
+import {spawnSync} from 'child_process'
+import {Command} from 'clipanion'
+import crypto from 'crypto'
+import os from 'os'
+import {parseTags} from '../../helpers/tags'
+import {apiConstructor} from './api'
+import {APIHelper, CIRCLECI, Provider, SUPPORTED_PROVIDERS} from './interfaces'
+
+// We use 127 as exit code for invalid commands since that is what *sh terminals return
+const BAD_COMMAND_EXIT_CODE = 127
+
+export class TraceCommand extends Command {
+  public static usage = Command.Usage({
+    description: 'Trace a command with a custom span and report it to Datadog.',
+    details: `
+            This command wraps another command, which it will launch, and report a custom span to Datadog.
+            See README for details.
+        `,
+    examples: [
+      [
+        'Trace a command with name "Say Hello" and report to Datadog',
+        'datadog-ci trace --name "Say Hello" -- echo "Hello World"',
+      ],
+      [
+        'Trace a command and report to the datadoghq.eu site',
+        'DATADOG_SITE=datadoghq.eu datadog-ci trace -- echo "Hello World"',
+      ],
+    ],
+  })
+  private command?: string[]
+
+  private config = {
+    apiKey: process.env.DATADOG_API_KEY || process.env.DD_API_KEY,
+    envVarTags: process.env.DD_TAGS,
+  }
+
+  private name?: string
+
+  private noFail?: boolean
+
+  public async execute() {
+    if (!this.command || !this.command.length) {
+      this.context.stderr.write('Missing command to run\n')
+
+      return 1
+    }
+
+    const [command, ...args] = this.command
+    const id = crypto.randomBytes(5).toString('hex')
+    const startTime = new Date().toISOString()
+    const spawnResult = spawnSync(command, args, {env: {...process.env, DD_CUSTOM_PARENT_ID: id}, stdio: 'inherit'})
```

Comment:
As your function is already asynchronous, could you `spawn` from within a promise?
As much as possible we use async functions to avoid blocking node's main thread. (Though I reckon code is a bit heavier to read/write)
```suggestion
    const childProcess = spawn(command, args, {env: {...process.env, DD_CUSTOM_PARENT_ID: id}, stdio: 'inherit'})
    const spawnResult = await new Promise((resolve, reject) => {
      childProcess.on('error', (error: Error) => {
        reject(error)
      })

      childProcess.on('close', (exitCode: number) => {
        resolve(exitCode)
      }
    })
```